### DISCUSSION TITLE
Is it possible to pick up search engine URL parameters for setting the default search engine?

### DISCUSSION THREAD

**Discussion by opusforlife2** [bromite/bromite#1968] [Ideas]
Search engines like DuckDuckGo let you save their settings as URL parameters (like `&kav=1` for infinite scroll), which is required when you have the browser configured for privacy, and can't keep cookies.

Is it possible to make it so that if I edit settings in DuckDuckGo and perform a search, Bromite recognises them as search engine parameters and offers to save _that_ as the default search engine?

For some reason Bromite/Chromium on Android is the only browser for which I have been unable to find a solution for this. Fenix and Desktop Chromium let you set the search engine URL yourself, while Desktop Firefox's limitation can be worked around by directly editing the mozlz4 search settings file with a mozlz4 editor.

**Comment by Disqu3-mirOir-qt** [comment]
with userscript easy way

**Reply by opusforlife2** [reply]
How does that work?

**Reply by ghost** [reply]
> How does that work?

When you search something, The Userscrpt would edit the url with parameters added.

**Reply by opusforlife2** [reply]
Ah. I see.

**Reply by opusforlife2** [reply]
If anyone finds or writes such a script, please post it in this thread.

**Comment by blogmangit** [comment]
> with userscript easy way

can u post the name of user script

**Reply by ghost** [reply]
> can u post the name of user script

If u have even little knowledge of javascript, you can make the Userscript yourself.

**Reply by blogmangit** [reply]
> > can u post the name of user script
> 
> If u have even little knowledge of javascript, you can make the Userscript yourself.

oh no sorry  man I don't  have any knowledge regarding anything I am that kind a person  who likes to get spoonfeed on everything. sorry again my bad that I disturbed you on this.

**Reply by kevin0t** [reply]
Yeah i want this script too. As i think chromium for android doesn't have the support for using a custom paramter, they all only use the basse domain, so the only was would be to use scripts.

**Comment by ncaat** [comment]
Here's a userscript that sets the DuckDuckGo URL parameters "k7" and "kj". Just change the parameters to whatever you want (in case you don't like the defaults :P)
```js
// ==UserScript==
// @name DuckDuckGo Params
// @match https://duckduckgo.com/*
// @run-at document-start
// ==/UserScript==
(()=>{
  var q=[['k7','00f'],['kj','0f0']] // set the params k7=00f and kj=0f0
    ,r=false,l=location,h=l.href,s=l.search,p=new URLSearchParams(s);
  q.forEach(n=>{
    if(!p.has(n[0])){
      r=true;
      p.set(...n);
    }});
  if(r) l.replace(h.replace(s.length>0?s:/$/,'?'+p.toString()));
})();
```

**Reply by kevin0t** [reply]
Thanks for the script bro ! though theme selection could be much better 😂.
Here my version of parameters 
```
// ==UserScript==
// @name DuckDuckGo Params
// @match https://duckduckgo.com/*
// @run-at document-start
// ==/UserScript==
(()=>{
  var q=[['kh','1'],['kg','p'],['kp','-2'],['kaf','1'],['kn',1'],['k1','-1'],['kl','us-en'] // set the params kh=1 and kg=p and kp=-2 and kaf=1 and kn=1 and k1=-1 and kl=us-en
    ,r=false,l=location,h=l.href,s=l.search,p=new URLSearchParams(s);
  q.forEach(n=>{
    if(!p.has(n[0])){
      r=true;
      p.set(...n);
    }});
  if(r) l.replace(h.replace(s.length>0?s:/$/,'?'+p.toString()));
})();

```

here kh=1 (https on) , kg=p(address bar-post) , kp=-2(safe search off) , kaf=1(full urls), kn=1(new window), k1=-1(advertisement off), kl=en(location -us)

**Reply by ncaat** [reply]
`['kn',1']` should probably be `['kn','1']`

**Reply by kevin0t** [reply]
yeah you are right , but anyway , new windows doesn't work even after correcting the script

**Reply by ncaat** [reply]
You also forgot the last `]` at the end. Version that works for me:
```js
// ==UserScript==
// @name DuckDuckGo Params
// @match https://duckduckgo.com/*
// @run-at document-start
// ==/UserScript==
(()=>{
  var q=[['kh','1'],['kg','p'],['kp','-2'],['kaf','1'],['kn','1'],['k1','-1'],['kl','us-en']] // set the params kh=1 and kg=p and kp=-2 and kaf=1 and kn=1 and k1=-1 and kl=us-en
    ,r=false,l=location,h=l.href,s=l.search,p=new URLSearchParams(s);
  q.forEach(n=>{
    if(!p.has(n[0])){
      r=true;
      p.set(...n);
    }});
  if(r) l.replace(h.replace(s.length>0?s:/$/,'?'+p.toString()));
})();
```

**Reply by kevin0t** [reply]
thanks it works now 👍. The new tab paramter combined with tab groups enabled in flags makes browsing super quick and accessible.

**Reply by kevin0t** [reply]
though i am thinking that using this specific search url parameter everytime would send a unique fingerprint to duckduckgo 😅. I hope many people use it with these custom setttings.

**Reply by opusforlife2** [reply]
Thanks, this is fantastic! Could DDG be replaced by Startpage here or is that a whole different ballgame?

**Reply by ncaat** [reply]
Yeah you just need to change the url.

**Reply by opusforlife2** [reply]
Holy wow. That's great. Thank you so much for this!

**Reply by ncaat** [reply]
No problem, happy to help.

**Reply by opusforlife2** [reply]
Hi again. I replaced the URL and parameters for Startpage, but it's not working. DDG worked perfectly, though!

**Reply by ncaat** [reply]
I could take a look at it if you post it here.

**Reply by opusforlife2** [reply]
```// @match https://www.startpage.com/*```

Startpage settings changes only give you one parameter.

```  var q=[['prfe','893eeb4b96e1e1abe8620e31e216606bdf308f98ac5feb5befbe1814e900d51bf1280e2af0161b3064b4b1fab92cdf6398863fa5cc5848c431b895e550b93ede3b1e20181cecf9da6f14bd90']] // set the params ```

The rest is the same. I tried using single brackets and double brackets around the parameter. Neither worked.

**Reply by ncaat** [reply]
It works for me. Are you sure everything else is the same?

**Reply by ncaat** [reply]
Here's the full working script:
```js
// ==UserScript==
// @name Startpage Params
// @match https://www.startpage.com/*
// @run-at document-start
// ==/UserScript==
(()=>{
  var q=[['prfe','893eeb4b96e1e1abe8620e31e216606bdf308f98ac5feb5befbe1814e900d51bf1280e2af0161b3064b4b1fab92cdf6398863fa5cc5848c431b895e550b93ede3b1e20181cecf9da6f14bd90']] // set the params
    ,r=false,l=location,h=l.href,s=l.search,p=new URLSearchParams(s);
  q.forEach(n=>{
    if(!p.has(n[0])){
      r=true;
      p.set(...n);
    }});
  if(r) l.replace(h.replace(s.length>0?s:/$/,'?'+p.toString()));
})();
```

**Reply by opusforlife2** [reply]
Ah, I hadn't realised we need to re-add the script to Bromite after editing the file. It works now. Thanks a lot once again!

**Comment by opusforlife2** [comment]
@csagan5 Could you please link @ncaat's comment on the search engine wiki page? It would be useful for users wanting custom search settings.

**Reply by csagan5** [reply]
Which version? There is one for StartPage and one for DuckDuckGo.

**Reply by opusforlife2** [reply]
The top comment? Readers could follow the comment chain and look at Startpage as well.

Or you could link both? It's up to you.
